### PR TITLE
Refactoring Charts.

### PR DESCRIPTION
**Pull request by Harikarthyk**
This PR Includes the following changes.

1. Chart Component moved to widgets.
2. All the variants are now managed through the dialog box.
3. switchTheme function is now added to the namespaces.

### CODE REVIEW

File: application/backend/src/main/java/io/typerefinery/websight/models/components/widgets/Chart.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    
+    private static final String DEFAULT_ID = "chart";
+    private static final String DEFAULT_CLASS_NAMES = "chart";
+    private static final String DEFAULT_MODULE = "chartComponent";
+
+    @Override
+    @PostConstruct
+    protected void init() {
+        this.id = DEFAULT_ID;
+        this.classNames = DEFAULT_CLASS_NAMES;
```
Comment (wildone) [inline application/backend/src/main/java/io/typerefinery/websight/models/components/widgets/Chart.java:34]: this is not required, please see existing classes.

File: application/backend/src/main/resources/apps/typerefinery/components/structure/page/clientlibs/functions.js [hunk 2]
```diff
@@ -3,9 +3,12 @@ window.Typerefinery.Components = Typerefinery.Components || {};
 window.Typerefinery.Theme = Typerefinery.Theme || {};
 window.MessageService = window.MessageService || {};
 window.MessageService.Client = MessageService.Client || {};
+window.Typerefinery.Components.Widgets = Typerefinery.Components.Widgets || {};
```
Comment (wildone) [inline application/backend/src/main/resources/apps/typerefinery/components/structure/page/clientlibs/functions.js:6]: this namespace does not match what the file is. This namespace should be Typerefinery.Components.Page meaning all functions associated with Page, so components could call these functions to get page level functionality.

File: application/backend/src/main/resources/apps/typerefinery/components/structure/page/clientlibs/functions.js [hunk 3]
```diff
@@ -3,9 +3,12 @@ window.Typerefinery.Components = Typerefinery.Components || {};
...
+window.Typerefinery.Components.Widgets.Charts = Typerefinery.Components.Widgets.Charts || {};
```
Comment (wildone) [inline application/backend/src/main/resources/apps/typerefinery/components/structure/page/clientlibs/functions.js:7]: all of this should be sitting in Charts client libs!

File: application/backend/src/main/resources/apps/typerefinery/components/structure/page/clientlibs/functions.js [hunk 4]
```diff
@@ -64,15 +67,9 @@ window.MessageService.Client = MessageService.Client || {};
...
-                            }else if ($component.getAttribute('data-module') === 'linechartComponent') {
-                                componentNs?.Graphs?.LineChart?.dataReceived(payload.data, $component)
-                            }else if ($component.getAttribute('data-module') === 'piechartComponent') {
-                                componentNs?.Graphs?.PieChart?.dataReceived(payload.data, $component)
-                            }else if ($component.getAttribute('data-module') === 'xaxisbarchartComponent') {
-                                componentNs?.Graphs?.XAxisBarChart?.dataReceived(payload.data, $component)
-                            }else if ($component.getAttribute('data-module') === 'barchartComponent') {
-                                componentNs?.Graphs?.BarChart?.dataReceived(payload.data, $component);
-                            }else if ($component.getAttribute('data-module') === 'tableComponent') {
+                            } else if ($component.getAttribute('data-module') === 'chartComponent') {
```
Comment (wildone) [inline application/backend/src/main/resources/apps/typerefinery/components/structure/page/clientlibs/functions.js:70]: each one of these should be in its own behaviour.js!

Comment (wildone) [reply application/backend/src/main/resources/apps/typerefinery/components/structure/page/clientlibs/functions.js:70]: we cant be updating some crazy monolith if block...

File: application/backend/src/main/resources/apps/typerefinery/components/structure/page/clientlibs/functions.js [hunk 5]
```diff
@@ -81,16 +78,37 @@ window.MessageService.Client = MessageService.Client || {};
...
+            }, 15);
+        }
+    }
+
     ns.init = () => {
         const rootElement = document.querySelector(':root');
         themeNs.rootElementStyle = getComputedStyle(rootElement);
         ns.handleTheme();
 
+        $("#themeHandler").click(function () {
```
Comment (wildone) [inline application/backend/src/main/resources/apps/typerefinery/components/structure/page/clientlibs/functions.js:103]: is themeHandler a namespace? you are searchin for all #themeHandler in whole page? eveything within init is local to a component!

File: application/backend/src/main/resources/apps/typerefinery/components/widgets/chart/clientlibs/behaviour.js [hunk 6]
```diff
@@ -0,0 +1,13 @@
+window.Typerefinery = window.Typerefinery || {};
+window.Typerefinery.Components = Typerefinery.Components || {};
+window.Typerefinery.Components.Widgets = Typerefinery.Components.Widgets || {};
+window.Typerefinery.Components.Widgets.Charts = Typerefinery.Components.Widgets.Charts || {};
+
+(function (ns, document, window) {
+  "use strict";
+  $(document).ready(function () {
+    $('[data-module="chartComponent"]').each(function () {
+      ns.init(this);
```
Comment (wildone) [inline application/backend/src/main/resources/apps/typerefinery/components/widgets/chart/clientlibs/behaviour.js:10]: this should include logic for 

```
   switch (componentConfig.chartVariant) {
      case "barChart":
        variantNs?.BarChart?.init($component)
```

File: application/backend/src/main/resources/apps/typerefinery/components/widgets/chart/clientlibs/functions.js [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+  };
+
+  ns.init = ($component) => { 
+    // parse json value from data-model attribute as component config.
+    const componentConfig = componentNs.getComponentConfig($component);
+   
+    // switch case for chart variants.
+    switch (componentConfig.chartVariant) {
+      case "barChart":
+        variantNs?.BarChart?.init($component);
```
Comment (wildone) [inline application/backend/src/main/resources/apps/typerefinery/components/widgets/chart/clientlibs/functions.js:38]: this logic should be in behaviour, no one will know that you are loading another init hidden within init. this should be simple.

File: application/backend/src/main/resources/apps/typerefinery/components/widgets/chart/clientlibs/variants/barchart/functions.js [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+window.Typerefinery.Components.Widgets = Typerefinery.Components.Widgets || {};
+window.Typerefinery.Components.Widgets.Charts = Typerefinery.Components.Widgets.Charts || {};
+window.Typerefinery.Components.Widgets.Charts.Variants = Typerefinery.Components.Widgets.Charts.Variants || {};
+window.Typerefinery.Components.Widgets.Charts.Variants.BarChart = Typerefinery.Components.Widgets.Charts.Variants.BarChart || {};
+window.Typerefinery.Components.Widgets.Charts.Instances = Typerefinery.Components.Widgets.Charts.Instances || {};
+
+(function (ns, typerefineryNs, componentNs, themeNs, chartInstanceNs, document, window) {
+    "use strict";
+
+    const DEFAULT_DATA = {
```
Comment (wildone) [inline application/backend/src/main/resources/apps/typerefinery/components/widgets/chart/clientlibs/variants/barchart/functions.js:13]: should this be added to namespace?

File: application/backend/src/main/resources/apps/typerefinery/components/widgets/chart/clientlibs/variants/barchart/functions.js [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+        };
+
+        chartInstanceNs[componentConfig.resourcePath].update();
+    };
+    ns.dataReceived = (data, $component) => {
+        // Passing {} because, The values from the model obj are fetched in bellow function definition.
+        ns.updateChartInstance(data, $component);
+    };
+
+    ns.init = ($component) => {
```
Comment (wildone) [inline application/backend/src/main/resources/apps/typerefinery/components/widgets/chart/clientlibs/variants/barchart/functions.js:206]: 👍

File: application/backend/src/main/resources/apps/typerefinery/components/widgets/chart/dialog/.content.json [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+  "tabs": {
+    "sling:resourceType": "wcm/dialogs/components/tabs",
+    "general": {
+      "sling:resourceType": "wcm/dialogs/components/tab",
+      "label": "General",
+      "chartVariant": {
+        "sling:resourceType": "wcm/dialogs/components/select",
+        "name": "chartVariant",
+        "label": "Choose Graph Variants",
+        "lineChart": {
```
Comment (wildone) [inline application/backend/src/main/resources/apps/typerefinery/components/widgets/chart/dialog/.content.json:12]: please add default option with no value, this way your defaults in the component should work.

File: application/backend/src/main/resources/apps/typerefinery/components/widgets/chart/dialog/.content.json [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+      "sling:resourceType": "wcm/dialogs/components/tab",
+      "label": "Data",
+      "dataSource": {
+        "sling:resourceType": "wcm/dialogs/components/textfield",
+        "name": "dataSource",
+        "description": "Please provide url to data source for this ticker.",
+        "label": "Data Source"
+      }
+    },
+    "websocketTab": {
```
Comment (wildone) [inline application/backend/src/main/resources/apps/typerefinery/components/widgets/chart/dialog/.content.json:55]: can you please update this to include /apps/typerefinery/components/dialog/websocket.

File: docker-compose.yml [hunk 12]
```diff
@@ -0,0 +1,38 @@
+version: "3.9"
```
Comment (wildone) [inline docker-compose.yml:1]: please remove this, we know this is not going to work as, none of our dependencies and configs are present.

**Review by wildone — CHANGES_REQUESTED** [review]

File: application/backend/src/main/resources/apps/typerefinery/components/structure/page/clientlibs/functions.js (hunk 4 above)
Comment (Harikarthyk) [reply application/backend/src/main/resources/apps/typerefinery/components/structure/page/clientlibs/functions.js:70]: Then whether the event listners should also be in the behaviour.js ?

image

File: application/backend/src/main/resources/apps/typerefinery/components/structure/page/clientlibs/functions.js (hunk 5 above)
Comment (Harikarthyk) [reply application/backend/src/main/resources/apps/typerefinery/components/structure/page/clientlibs/functions.js:103]: Here we are just adding the click event listener to that id.

File: application/backend/src/main/resources/apps/typerefinery/components/widgets/chart/clientlibs/functions.js (hunk 7 above)
Comment (Harikarthyk) [reply application/backend/src/main/resources/apps/typerefinery/components/widgets/chart/clientlibs/functions.js:38]: Ok, will update this.

File: application/backend/src/main/resources/apps/typerefinery/components/widgets/chart/clientlibs/variants/barchart/functions.js (hunk 8 above)
Comment (Harikarthyk) [reply application/backend/src/main/resources/apps/typerefinery/components/widgets/chart/clientlibs/variants/barchart/functions.js:13]: If I add this out side the name space and pass as an arg we can use the same name in all the functions.js

Got "DEFAULT_DATA Already used"

File: application/backend/src/main/resources/apps/typerefinery/components/widgets/chart/dialog/.content.json (hunk 10 above)
Comment (Harikarthyk) [reply application/backend/src/main/resources/apps/typerefinery/components/widgets/chart/dialog/.content.json:12]: Ok. It is already handled in code, will remove from here.

File: application/backend/src/main/resources/apps/typerefinery/components/widgets/chart/dialog/.content.json (hunk 11 above)
Comment (Harikarthyk) [reply application/backend/src/main/resources/apps/typerefinery/components/widgets/chart/dialog/.content.json:55]: Ok, sure.

File: docker-compose.yml (hunk 12 above)
Comment (Harikarthyk) [reply docker-compose.yml:1]: Yup, will do, sorry by mistake this got added, I thought of removing this before the scrum call.

File: application/backend/src/main/resources/apps/typerefinery/components/widgets/chart/clientlibs/behaviour.js (hunk 6 above)
Comment (Harikarthyk) [reply application/backend/src/main/resources/apps/typerefinery/components/widgets/chart/clientlibs/behaviour.js:10]: Ok.

**Pull request merged**